plugins/Gravatar/Gravatar.php

```diff
@@ -104,24 +104,14 @@ function config() {
 	 * Register event hooks for plugin.
 	 */
 	function hooks() {
-		return array(
-			'EVENT_USER_AVATAR' => 'user_get_avatar',
-			'EVENT_LAYOUT_RESOURCES' => 'csp_headers',
-		);
-	}
-
-	/**
-	 * Add Content-Security-Policy for retrieving Avatar images.
-	 *
-	 * @return void
-	 */
-	function csp_headers() {
-		# Policy for images: Allow gravatar URL
 		if( config_get( 'show_avatar' ) !== OFF ) {
 			# Set CSP header
-			header( "Content-Security-Policy: img-src 'self' " .
-				self::getAvatarUrl() );
+			http_csp_add( 'img-src', self::getAvatarUrl() );
```

Comment:
I would recommend to keep the `csp_headers()` function as a separate event hook, like it was before. I understand we have a problem of sequence in the core, i.e. by the time *EVENT_LAYOUT_RESOURCES* is signaled `http_security_headers()` has already been called, but I don't think it's conceptually correct for the `hooks()` method itself to update the CSP.

To be clean, I think we should declare a new event to allow plugins to properly hook their code. I can think of 2 approaches here:

1. Use a CSP-specific event (e.g. *EVENT_CSP_INIT* or something similar), to be signaled from `http_security_headers()` before issuing the actual header; simple, but not reusable for other purposes.

2. Go for a more generic approach, i.e. an event allowing plugins to add other headers. Here the challenge is to decide where to signal the event... For CSP it must occur *before* we send headers (i.e. at beginning of the if statement in `http_all_headers()`), but then plugins won't be able to override Mantis default headers (assuming we want to let them do it) reducing the event's . In that case maybe we would need 2 events, like *EVENT_HTTP_HEADERS_PREPARE* *EVENT_HTTP_HEADERS_SENT* ?

My preference would be for 1.